spec/avromatic/model/raw_serialization_spec.rb

```diff
@@ -26,47 +27,125 @@
       expect(instance.value_attributes_for_avro).to eq(expected)
     end
 
+    context "with a nested record" do
+      let!(:test_class) do
+        Avromatic::Model.model(value_schema_name: 'test.nested_record')
+      end
+      let(:sub) { Avromatic.nested_models['test.__nested_record_sub_record'].new(str: 'b', i: 1) }
+      let(:values) { { str: 'a', sub: sub } }
+
+      it "reuses cacheable attributes" do
+        expected = values.deep_stringify_keys
+        expected['sub'] = { encoding_provider => sub }
+        actual = instance.value_attributes_for_avro
+        expect(actual).to eq(expected)
+        expect(actual['sub'][encoding_provider].object_id).to eq(sub.object_id)
+      end
+    end
+
+    context "with repeated references to a named type" do
+      let!(:test_class) do
+        Avromatic::Model.model(value_schema_name: 'test.wrapper')
+      end
+      let(:wrapped1) { Avromatic.nested_models['test.wrapped1'].new(i: 42) }
+      let(:wrapped2) { Avromatic.nested_models['test.wrapped2'].new(i: 78) }
+      let(:values) { { sub1: wrapped1, sub2: wrapped1, sub3: wrapped2 } }
+
+      it "reuses cacheable attributes" do
+        expected = values.deep_stringify_keys.map { |k, v| [k, { encoding_provider => v }] }.to_h
```

Comment:
Make sense. Minor, but this might be slightly cleaner than going from hash to array to hash:

```ruby
values.deep_stringify_keys.each_with_object({}) { |(k, v), hash| hash[k] = { encoding_provider => v } }
```